tests/conftest.py

```diff
@@ -58,22 +60,78 @@ def configure_epics_environment():
     os.environ["EPICS_PVA_AUTO_ADDR_LIST"] = "NO"
 
 
+_ALLOWED_CORO_TASKS = {"async_finalizer", "async_setup", "async_teardown"}
+
+
+@pytest.fixture(autouse=True, scope="function")
+def assert_no_pending_tasks(request: FixtureRequest):
+    """
+    if "RE" in request.fixturenames:
+        # The RE fixture will do the same, as well as some additional loop cleanup
+        return
+    """
+
+    # There should be no tasks pending after a test has finished
+    fail_count = request.session.testsfailed
+
+    def error_and_kill_pending_tasks():
+        loop = asyncio.get_event_loop()
+        unfinished_tasks = {
+            task
+            for task in asyncio.all_tasks(loop)
+            if task.get_coro().__name__ not in _ALLOWED_CORO_TASKS and not task.done()
+        }
+        for task in unfinished_tasks:
+            task.cancel()
+
+        # If the tasks are still pending because the test failed
+        # for other reasons then we don't have to error here
+        if unfinished_tasks and request.session.testsfailed == fail_count:
+            raise RuntimeError(
+                f"Not all tasks closed during test {request.node.name}:\n"
+                f"{pprint.pformat(unfinished_tasks)}"
+            )
```

Comment:
This is a duplicated error message from the one in the RE fixture. Come to think of it, I'm confused as to why we need this other finalizer, is to to catch tasks hanging around when not using the RE?

Either way, can we abstract the duplicated code into another function please?

Reply:
My reasoning was that non-RE fixture is getting the active event loop which will be different from the one the RE fixture is runnning.

I'll look into getting the RE fixture's event loop again.

Reply:
Okay the problem is that if we close the event-loop in `RE` then there's no way to get it or unhandled tasks within it from inside the finalizer of `assert_no_pending_tasks` which needs to run after RE has finished. We'd need to store it globally.

I think there also might be problems with cancelling the tasks after the bluesky event-loop thread has been joined. I don't think there's a convenient way to get the loop that was passed into the `RunEngine`.

Unless you have any ideas on how to handle it, I'd say it makes sense to fail in different places for the different event-loops.

Reply:
I think it makes sense to fail in different places, but I think the common code from the two functions could be pulled out into a utility function that both call...